src/waypoint.jsx

```diff
@@ -333,13 +352,30 @@ export default class Waypoint extends React.Component {
    * @return {Object}
    */
   render() {
+    if (this.props.noWrapper) {
+      const child = React.Children.only(this.props.children);
+      const ref = (node) => {
+        this.refElement(node);
+        if (this.props.children.ref) {
+          this.props.children.ref(node);
+        }
+      };
+      return React.cloneElement(child, { ref });
+    }
+
+    if (this.props.children) {
+      return <div ref={this.refElement}>{this.props.children}</div>;
+    }
+
     // We need an element that we can locate in the DOM to determine where it is
     // rendered relative to the top of its context.
     return <span ref={this.refElement} style={{ fontSize: 0 }} />;
   }
 }
 
 Waypoint.propTypes = {
+  children: PropTypes.oneOfType([PropTypes.node, PropTypes.array]),
+  noWrapper: PropTypes.bool,
```

Comment:
I don't think it is a good idea to conditionally add a div if there are multiple children or of the children is a functional component because that would be a pretty surprising and unexpected thing for folks. I think it would be much better to provide a useful error message and always require an only child that is not a functional component. That is clear, consistent, and flexible.